pkg/types/ccipocr3/v2/common_types.go

```diff
@@ -0,0 +1,176 @@
+package ccipocr3
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"strings"
+)
+
+// UnknownAddress represents a raw address with an unknown encoding.
+type UnknownAddress []byte
+
+// NewUnknownAddressFromHex creates a new UnknownAddress from a hex string.
+func NewUnknownAddressFromHex(s string) (UnknownAddress, error) {
+	b, err := NewBytesFromString(s)
+	if err != nil {
+		return nil, err
+	}
+	return UnknownAddress(b), nil
+}
+
+// String returns the hex representation of the unknown address.
+func (a UnknownAddress) String() string {
+	return Bytes(a).String()
+}
+
+func (a UnknownAddress) MarshalJSON() ([]byte, error) {
+	return Bytes(a).MarshalJSON()
+}
+
+func (a *UnknownAddress) UnmarshalJSON(data []byte) error {
+	return (*Bytes)(a).UnmarshalJSON(data)
+}
+
+// UnknownEncodedAddress represents an encoded address with an unknown encoding.
+type UnknownEncodedAddress string
+
+type Bytes []byte
+
+func NewBytesFromString(s string) (Bytes, error) {
+	if len(s) < 2 {
+		return nil, fmt.Errorf("Bytes must be of at least length 2 (i.e, '0x' prefix): %s", s)
+	}
+
+	if !strings.HasPrefix(s, "0x") {
+		return nil, fmt.Errorf("Bytes must start with '0x' prefix: %s", s)
+	}
+
+	b, err := hex.DecodeString(s[2:])
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode hex: %w", err)
+	}
+
+	return Bytes(b), nil
+}
+
+func (b Bytes) String() string {
+	return "0x" + hex.EncodeToString(b)
+}
+
+func (b Bytes) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf(`"%s"`, b.String())), nil
+}
+
+func (b *Bytes) UnmarshalJSON(data []byte) error {
+	v := string(data)
+	if len(v) < 2 {
+		return fmt.Errorf("Bytes must be of at least length 2 (i.e, '0x' prefix): %s", v)
+	}
+
+	// trim the start and end double quotes
+	v = v[1 : len(v)-1]
+
+	if !strings.HasPrefix(v, "0x") {
+		return fmt.Errorf("Bytes must start with '0x' prefix: %s", v)
+	}
+
+	// Decode everything after the '0x' prefix.
+	bs, err := hex.DecodeString(v[2:])
+	if err != nil {
+		return fmt.Errorf("failed to decode hex: %w", err)
+	}
+
+	*b = bs
+	return nil
+}
+
+type Bytes32 [32]byte
+
+func NewBytes32FromString(s string) (Bytes32, error) {
+	if len(s) < 2 {
+		return Bytes32{}, fmt.Errorf("Bytes32 must be of at least length 2 (i.e, '0x' prefix): %s", s)
+	}
+
+	if !strings.HasPrefix(s, "0x") {
+		return Bytes32{}, fmt.Errorf("Bytes32 must start with '0x' prefix: %s", s)
+	}
+
+	b, err := hex.DecodeString(s[2:])
+	if err != nil {
+		return Bytes32{}, fmt.Errorf("failed to decode hex: %w", err)
+	}
+
+	var res Bytes32
+	copy(res[:], b)
```

Comment:
Would we want to put the bytes at the end of the array? 
ie. should the result of the conversion be `"0x1234" -> "0x000...01234"` rather than `"0x1234" -> "0x12340000..."`?

Reply:
The former seems to be more of what I would expect, at least if the hex bytes represented a number

Reply:
This code isn't changed from v1. There aren't many places using `Bytes32` after this change, we would have to review those to see if it makes sense to modify the behavior.

I guess my expectation would be `s == NewBytes32FromString(s).String()` which definitely isn't the case with the current implementation which also silently truncates the input.